lib/urn.js

```diff
@@ -149,31 +155,28 @@ module.exports = {
 
       if (schemaLength <= 0) {
         throw new Error('Wrong fomatting');
-      }
-
-      if (schemaLength > 255) {
+      } else if (schemaLength > 255) {
         throw new Error('Maximum length of URN schema is 255');
       }
 
-      // Each hex char is 4 bits
-      var idLength = (identity.length - schemaLength - 1) / 2;
-
-      if (idLength != idLength.toFixed()) {
-        throw new Error('Hex chars are of odd count');
-      }
+      var schema = identity.slice(0, schemaLength);
+      var id = identity.slice(schemaLength + 1);
+      var schemaHex = stringToHex(schema);
+      var idHex = RECOGNIZED_PROTOCOLS.has(schema) ? id : stringToHex(id);
+      var idLength = idHex.length / 2;
 
       if (idLength <= 0) {
         throw new Error('ID is empty');
+      } else if (idLength > MAX_HEX_ID_LENGTH) {
+        throw new Error(`Maximum length of ID is ${MAX_HEX_ID_LENGTH}`);
+      } else if (idLength != idLength.toFixed()) {
+        throw new Error('Hex chars are of odd count');
```

Comment:
Optional: Another "extra" foolproof check could be `schemaHex.length / 2 !==  schemaLength`
Since `stringToHex` could misbehave and number of bytes in `schemaHex` doesn't match number of symbols in `schema`